File: crypto/VSS.py
from decimal import Decimal, getcontext
import random

# TypeScript の Decimal.js での精度設定に対応（必要に応じて調整）
getcontext().prec = 10_000_000

def divmod_decimal(a, b, n):
    """
    拡張ユークリッド互除法を用いて (a * b^(-1)) mod n を計算する関数です。
    b の逆元が存在しない場合は 0 を返します。
    """
    a = Decimal(a)
    b = Decimal(b)
    n = Decimal(n)
    t = Decimal(0)
    nt = Decimal(1)
    r = n
    nr = b % n
    while nr != 0:
        quot = r // nr
        t, nt = nt, t - quot * nt
        r, nr = nr, r - quot * nr
    if r > 1:
        return Decimal(0)
    if t < 0:
        t = t + n
    return (a * t) % n

def expmod(x, n, mod):
    """
    二分累乗法を用いて (x ** n) mod mod を計算する関数です。
    """
    x = Decimal(x)
    n = Decimal(n)
    mod = Decimal(mod)
    ret = Decimal(1)
    while n > 0:
        if n % 2 != 0:
            ret = (ret * x) % mod
        x = (x * x) % mod
        n = n // 2
    return ret

def power(x, n):
    """
    二分累乗法による単純な累乗計算 (x ** n) を行う関数です。
    """
    x = Decimal(x)
    n = Decimal(n)
    ret = Decimal(1)
    while n > 0:
        if n % 2 != 0:
            ret = ret * x
        x = x * x
        n = n // 2
    return ret

def random_decimal(lower, upper):
    """
    lower 以上 upper 未満のランダムな整数（Decimal）を返します。
    """
    lower = Decimal(lower)
    upper = Decimal(upper)
    if lower > upper:
        lower, upper = upper, lower
    return Decimal(random.randrange(int(lower), int(upper)))

def polynomial(x, a, q):
    """
    係数リスト a を持つ多項式を、x において q で割った余りを返します。
    a[0] が秘密となり、残りはランダム係数です。
    """
    value = a[0]
    for i in range(1, len(a)):
        # x^i mod q を expmod で計算し、その後係数をかけたものを加算
        value = (value + expmod(x, Decimal(i), q) * a[i]) % q
    return value

def split(secret, total_shares, threshold, modulus_p, modulus_q, generator):
    """
    Feldman の VSS による秘密分散を行う関数
    
    Parameters:
      secret: 分散する秘密（整数または16進数文字列）
      total_shares: 生成するシェアの総数
      threshold: 復元に必要な最小シェア数
      modulus_p: コミットメント用の素数
      modulus_q: 秘密分散用の素数（秘密はこの値未満である必要あり）
      generator: 生成元の値
    """
    # 秘密を整数に変換
    if isinstance(secret, str) and secret.startswith("0x"):
        secret_int = Decimal(int(secret, 16))
    else:
        secret_int = Decimal(int(secret))
    
    if secret_int > modulus_q:
        raise ValueError(f"秘密の整数値({secret_int})は modulus_q({modulus_q}) より小さい必要があります")

    coefficients = [secret_int]
    shares_list = []
    commitments = [{"index": Decimal(0), "value": expmod(generator, secret_int, modulus_p)}]

    for i in range(1, threshold):
        random_coeff = random_decimal(Decimal(1), modulus_q - Decimal(1))
        coefficients.append(random_coeff)
        commitments.append({"index": Decimal(i), "value": expmod(generator, random_coeff, modulus_p)})

    for share_index in range(total_shares):
        x_value = Decimal(share_index + 1)
        share = {"x": x_value, "y": polynomial(x_value, coefficients, modulus_q)}
        shares_list.append(share)

    return {"shares": shares_list, "commitments": commitments}



def lagrange_basis(data, j, q):
    """
    シェアリスト data からインデックス j のシェアに対するラグランジュ基底多項式の
    分子と分母を計算します。
    """
    numerator = Decimal(1)
    denominator = Decimal(1)
    x_j = data[j]["x"]
    
    for i in range(len(data)):
        if i != j:  # 自分自身は除外
            x_i = data[i]["x"]
            numerator = (numerator * x_i) % q
            denominator = (denominator * (x_i - x_j)) % q
    
    return numerator, denominator

def lagrange_interpolate(data, q):
    """
    ラグランジュ補間を用いてシェアから秘密を再構成します。
    f(0) = ∑(y_j * ∏(x_i/(x_i-x_j))) を計算します。
    """
    secret = Decimal(0)
    
    for j in range(len(data)):
        num, den = lagrange_basis(data, j, q)
        # L_j(0) = ∏(x_i)/(∏(x_i-x_j))
        L_j_0 = divmod_decimal(num, den, q)
        term = (data[j]["y"] * L_j_0) % q
        secret = (secret + term) % q
    
    return secret

def combine(shares, modulus_q, return_string=True):
    """
    シェアから秘密を復元します。
    
    Parameters:
      shares: シェアのリスト
      modulus_q: 秘密分散で使用した素数
      return_string: 文字列として返すかどうか
                     (True の場合、元の秘密が 10 進数文字列であればその形に合わせて返します)
    """
    decimal_shares = [{"x": Decimal(share["x"]), "y": Decimal(share["y"])} for share in shares]
    recovered_secret_int = lagrange_interpolate(decimal_shares, modulus_q)
    
    if return_string:
        # 10進数文字列として返す（hex() ではなく単に文字列化する）
        return str(int(recovered_secret_int))
    else:
        return int(recovered_secret_int)

def verify(share, C, prime, generator, q):
    """
    与えられたシェアがコミットメント C と整合するか検証します。
    
    シェア (x, y) に対して、検証すべき等式は  
      g^y ≡ ∏_{i=0}^{k-1} (C_i)^(x^i mod q) (mod prime)
    となります。
    
    Parameters:
      share: {'x': Decimal, 'y': Decimal}
      C: コミットメントのリスト。各要素は {'index': Decimal, 'value': Decimal}
      prime: p（Decimal）
      generator: g（Decimal）
      q: シークレット分散で用いた素数（Decimal）
    """
    p = Decimal(prime)
    g = Decimal(generator)
    lG = expmod(g, share["y"], p)
    rG = Decimal(1)
    for i in range(len(C)):
        commitment = C[i]
        e = power(share["x"], Decimal(i)) % q
        basis = expmod(commitment["value"], e, p)
        rG = (rG * basis) % p
    return rG == lG

# --- 以下は動作確認用のテストコード ---
if __name__ == "__main__":
    from parameters import KEY_1024_PARAMS

    # 高精度計算のための設定（テスト用に精度を調整）
    getcontext().prec = 1000

    # 暗号パラメータの読み込み
    MODULUS_P = Decimal(KEY_1024_PARAMS["p"])
    MODULUS_Q = Decimal(KEY_1024_PARAMS["q"])
    GENERATOR = Decimal(KEY_1024_PARAMS["g"])

    # テスト設定
    TOTAL_SHARES = 5
    THRESHOLD = 3

    print("\n=== 秘密分散テスト ===")
    secret_int = "123"  # テスト用秘密（10進数として扱う）
    print(f"元の秘密: {secret_int}")
    
    # 秘密を分散
    vss_result = split(secret_int, TOTAL_SHARES, THRESHOLD, MODULUS_P, MODULUS_Q, GENERATOR)
    shares = vss_result["shares"]
    commitments = vss_result["commitments"]

    print("\nShares:")
    for i, share in enumerate(shares):
        print(f"Share {i+1}: x={share['x']}, y={share['y']}")
    
    print("\nCommitments:")
    for i, commitment in enumerate(commitments):
        print(f"C_{i}: {commitment['value']}")

    # 最初の threshold 個のシェアから秘密を再構成
    test_shares = shares[:THRESHOLD]
    recovered_secret = combine(test_shares, MODULUS_Q, return_string=True)
    print(f"\n最初の{THRESHOLD}個のシェアから復元した秘密: {recovered_secret}")
    print(f"元の秘密と一致: {recovered_secret == secret_int}")

    # 別の threshold 個のシェアから秘密を再構成
    test_shares2 = [shares[0], shares[2], shares[4]]
    recovered_secret2 = combine(test_shares2, MODULUS_Q, return_string=True)
    print(f"\n別の{THRESHOLD}個のシェアから復元した秘密: {recovered_secret2}")
    print(f"元の秘密と一致: {recovered_secret2 == secret_int}")

    # 各シェアのコミットメント検証
    for i, share in enumerate(shares):
        valid = verify(share, commitments, MODULUS_P, GENERATOR, MODULUS_Q)
        print(f"シェア {i+1} の検証結果: {valid}")

File: crypto/VSS.ts
import { Decimal } from "decimal.js";
// rounding
// Decimal.set({ crypto: true }); // 予測不可能なランダム生成
Decimal.set({ precision: 1e7 }); // 小数点
Decimal.set({ toExpPos: 0 }); //指数部を指定
Decimal.set({ modulo: Decimal.ROUND_FLOOR }); // mod 記号の指定
Decimal.set({ rounding: 5 }); //丸目モードの指定、

// DSA key generator
// from cryptography.hazmat.primitives.asymmetric by python3
// https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf

function divmod(a: Decimal, b: Decimal, n: Decimal) {
	let aCopy = Decimal.isDecimal(a) ? a : new Decimal(a);
	let bCopy = Decimal.isDecimal(b) ? b : new Decimal(b);
	let nCopy = Decimal.isDecimal(n) ? n : new Decimal(n);
	let t = new Decimal("0");
	let nt = new Decimal("1");
	let r = nCopy;
	let nr = bCopy.mod(n);
	let tmp;
	while (!nr.isZero()) {
		let quot = Decimal.floor(r.div(nr));
		tmp = nt;
		nt = t.sub(quot.times(nt));
		t = tmp;
		tmp = nr;
		nr = r.sub(quot.times(nr));
		r = tmp;
	}
	if (r.greaterThan(1)) return new Decimal(0);
	if (t.isNegative()) t = t.add(n);
	return aCopy.times(t).mod(n);
}

function expmod(x: Decimal, n: Decimal, mod: Decimal): Decimal {
	let ret = new Decimal(1);
	while (n.greaterThan(0)) {
		if (!n.mod(2).equals(0)) ret = ret.mul(x).mod(mod); // n の最下位bitが 1 ならば x^(2^i) をかける
		x = x.mul(x).mod(mod);
		n = n.div(2).floor(); // n を1bit 左にずらす
	}
	return ret;
}

function pow(x: Decimal, n: Decimal): Decimal {
	let ret = new Decimal(1);
	while (n.greaterThan(0)) {
		if (!n.mod(2).equals(0)) ret = ret.mul(x); // n の最下位bitが 1 ならば x^(2^i) をかける
		x = x.mul(x);
		n = n.div(2).floor(); // n を1bit 左にずらす
	}
	return ret;
}

export interface DecimalShare {
	x: Decimal;
	y: Decimal;
}
export interface DecimalCommitment {
	i: Decimal;
	c: Decimal;
}

function random(lower: Decimal, upper: Decimal) {
	if (lower > upper) {
		const temp = lower;
		lower = upper;
		upper = temp;
	}

	return lower.add(
		Decimal.random()
			.times(upper.sub(lower.add(1)))
			.floor()
	);
}

// Polynomial function where `a` is the coefficients
function polynomial(x: Decimal, a: Decimal[], q: Decimal) {
	let value = a[0];
	for (let i = 1; i < a.length; i++) {
		// value = value.add(x.pow(i).times(a[i]));
		value = value.add(expmod(x, new Decimal(i), q).times(a[i])).mod(q);
	}
	return value;
}

function split(
	secret: string,
	n: number,
	k: number,
	p: Decimal,
	q: Decimal,
	generator: Decimal
) {
	if (
		Number.isInteger(secret) ||
		Number.isInteger(p) ||
		Number.isInteger(generator)
	) {
		throw new TypeError(
			"The shamir.split() function must be called with a String<secret>" +
				"but got Number<secret>."
		);
	}

	if (Number.isInteger(p)) {
		throw new TypeError(
			"The shamir.split() function must be called with a Decimal<prime>" +
				"but got Number<prime>."
		);
	}

	if (secret.substring(0, 2) !== "0x") {
		throw new TypeError(
			"The shamir.split() function must be called with a" + "Decimal<secret>"
		);
	}

	if (!Decimal.isDecimal(p)) {
		throw new TypeError(
			"The shamir.split() function must be called with a" + "Decimal<prime>"
		);
	}

	if (!Decimal.isDecimal(generator)) {
		throw new TypeError(
			"The shamir.split() function must be called with a" + "Decimal<generator>"
		);
	}

	const S = new Decimal(secret);
	const g = new Decimal(generator);

	if (S.greaterThan(q)) {
		throw new RangeError(
			"The String<secret> must be less than the Decimal<prime>."
		);
	}

	if (g.greaterThan(p)) {
		throw new RangeError(
			"The Decimal<generator> must be less than the Decimal<prime>."
		);
	}

	let a: Decimal[] = [S];
	let D: DecimalShare[] = [];
	let C: DecimalCommitment[] = [
		{
			i: new Decimal("0"),
			// c: g.pow(S).mod(p)
			c: expmod(g, S, p),
		},
	];

	for (let i = 1; i < k; i++) {
		let coeff = random(new Decimal(1), q.sub(1));
		a.push(coeff);
		C.push({
			i: new Decimal(i),
			c: expmod(g, coeff, p),
		});
	}
	for (let i = 0; i < n; i++) {
		let x = new Decimal(i + 1);
		D.push({
			x,
			y: polynomial(x, a, q),
		});
	}
	return { D, C };
}

function lagrangeBasis(data: DecimalShare[], j: number, q: Decimal) {
	// Lagrange basis evaluated at 0, i.e. L(0).
	// You don't need to interpolate the whole polynomial to get the secret, you
	// only need the constant term.
	let denominator = new Decimal(1);
	let numerator = new Decimal(1);
	for (let i = 0; i < data.length; i++) {
		if (!data[j].x.equals(data[i].x)) {
			denominator = denominator.times(data[i].x.minus(data[j].x)).mod(q);
		}
	}

	for (let i = 0; i < data.length; i++) {
		if (!data[j].x.equals(data[i].x)) {
			numerator = numerator.times(data[i].x).mod(q);
		}
	}

	return {
		numerator,
		denominator,
	};
}

function lagrangeInterpolate(data: DecimalShare[], q: Decimal) {
	let S = new Decimal(0);

	for (let i = 0; i < data.length; i++) {
		let basis = lagrangeBasis(data, i, q);
		S = S.add(
			data[i].y.times(divmod(basis.numerator, basis.denominator, q))
		).mod(q);
	}

	// const rest = S.mod(q);
	const rest = S;

	return rest;
}

function combine(shares: DecimalShare[], prime: Decimal) {
	const p = prime;

	// Wrap with Decimal on the input shares
	const decimalShares = shares.map((share) => ({
		x: new Decimal(share.x),
		y: new Decimal(share.y),
	}));

	return lagrangeInterpolate(decimalShares, p);
}

function verify(
	share: DecimalShare,
	C: DecimalCommitment[],
	prime: Decimal,
	generator: Decimal
) {
	const p = new Decimal(prime);
	const g = new Decimal(generator);
	let lG = expmod(g, new Decimal(share.y), p);
	let rG = new Decimal(1);
	console.log("init rG: ", rG.toHex());
	for (let i = 0; i < C.length; i++) {
		console.log("++++++++ com " + i);
		let commitment = C[i];
		let e = pow(share.x, new Decimal(i));
		let basis = expmod(commitment.c, e, p);
		rG = rG.mul(basis).mod(p);
		console.log("share.x: ", share.x.toHex());
		console.log("share.y: ", share.y.toHex());
		console.log("commitment: ", commitment.c.toHex());
		console.log("e: ", e.toHex());
		console.log("basis: ", basis.toHex());
		console.log("c rg: ", rG.toHex());
	}
	console.log("rg: " + rG.toHex());
	console.log("lg: " + lG.toHex());
	return rG.equals(lG);
}

export default { split, combine, verify, pow, expmod, divmod, random };
